types/tablet_v2/wlr_tablet_v2.c

```diff
@@ -0,0 +1,309 @@
+		add_tablet_tool_client(seat_client, tool_pos);
+	}
+}
+
+static struct zwp_tablet_manager_v2_interface manager_impl = {
+	.get_tablet_seat = get_tablet_seat,
+	.destroy = tablet_manager_destroy,
+};
+
+static struct wlr_tablet_manager_client_v2 *tablet_manager_client_from_resource (
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_tablet_manager_v2_interface,
+		&manager_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void wlr_tablet_manager_v2_destroy(struct wl_resource *resource) {
+	struct wlr_tablet_manager_client_v2 *client = tablet_manager_client_from_resource(resource);
+	if (!client) {
+		return;
+	}
+
+	struct wlr_tablet_seat_client_v2 *pos;
+	struct wlr_tablet_seat_client_v2 *tmp;
+	wl_list_for_each_safe(pos, tmp, &client->tablet_seats, client_link) {
+		wlr_tablet_seat_client_v2_destroy(pos->resource);
+	}
+
+	wl_list_remove(&client->link);
+	//wl_list_remove(&client->client_destroy.link);
+
+	free(client);
+	wl_resource_set_user_data(resource, NULL);
+}
+
+static void tablet_v2_bind(struct wl_client *wl_client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_tablet_manager_v2 *manager = data;
+	assert(wl_client && manager);
+
+	struct wlr_tablet_manager_client_v2 *client =
+		calloc(1, sizeof(struct wlr_tablet_manager_client_v2));
+	if (client == NULL) {
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+
+	wl_list_init(&client->tablet_seats);
+
+	client->resource =
+		wl_resource_create(wl_client, &zwp_tablet_manager_v2_interface, version, id);
+	if (client->resource == NULL) {
+		free(client);
+		wl_client_post_no_memory(wl_client);
+		return;
+	}
+	client->client = wl_client;
+	client->manager = manager;
+
+	wl_resource_set_implementation(client->resource, &manager_impl, client,
+		wlr_tablet_manager_v2_destroy);
+	wl_list_insert(&manager->clients, &client->link);
+}
+
+static void handle_display_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_tablet_manager_v2 *tablet =
+		wl_container_of(listener, tablet, display_destroy);
+	wlr_tablet_v2_destroy(tablet);
+}
+
+void wlr_tablet_v2_destroy(struct wlr_tablet_manager_v2 *manager) {
+	struct wlr_tablet_manager_client_v2 *tmp;
+	struct wlr_tablet_manager_client_v2 *pos;
+
+	wl_list_for_each_safe(pos, tmp, &manager->clients, link) {
+		wlr_tablet_manager_v2_destroy(pos->resource);
+	}
+
+	wl_global_destroy(manager->wl_global);
+	free(manager);
+}
+
+struct wlr_tablet_manager_v2 *wlr_tablet_v2_create(struct wl_display *display) {
+	struct wlr_tablet_manager_v2 *tablet =
+		calloc(1, sizeof(struct wlr_tablet_manager_v2));
+	if (!tablet) {
+		return NULL;
+	}
+
+	wl_list_init(&tablet->clients);
+	wl_list_init(&tablet->seats);
+
+	tablet->display_destroy.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &tablet->display_destroy);
+
+	tablet->wl_global = wl_global_create(display,
+		&zwp_tablet_manager_v2_interface, 1, tablet, tablet_v2_bind);
```

Comment:
Can you define `TABLET_MANAGER_VERSION` at the beginning of the file, so that it's obvious which version we implement?